signedurlmanager.go

```diff
@@ -0,0 +1,65 @@
+package main
+
+import (
+	"github.com/taskcluster/taskcluster-client-go/queue"
+	"log"
+	"os"
+	"strconv"
+	"time"
+)
+
+// This function is called in a dedicated go routine to both serve signed urls
+// and to update them before they expire
+func SignedURLsManager() {
+	// prematurity specifies the number of seconds prior to expiry that new
+	// signed urls should be fetched, in order that stale credentials are not
+	// used. Should be at least a few seconds.
+	prematurity := os.Getenv("REFRESH_URLS_PREMATURELY_SECS")
+	// convert number of seconds to an integer
+	premInt, err := strconv.Atoi(prematurity)
+	if err != nil {
+		log.Printf("Environment variable REFRESH_URLS_PREMATURELY_SECS should be an integer number of seconds, but is '%v'.", prematurity)
+		log.Fatalf("This variable represents the number of seconds before signed URLs expire, that they should be refreshed.")
+	}
+	// signedURLs is the variable where we store the current valid signed urls
+	var signedURLs *queue.PollTaskUrlsResponse
+	// updateMe is a channel to send a message to when we need to update signed
+	// urls because either we don't have any yet (i.e. first time) or they are
+	// about to expire...
+	updateMe := make(<-chan time.Time)
+	// function to refresh signed urls
+	updateUrls := func() {
+		// When a worker wants to poll for pending tasks it must call
+		// `queue.pollTaskUrls(provisionerId, workerType)` which then returns
+		// an array of objects on the form `{signedPollUrl, signedDeleteUrl}`.
+		signedURLs, _ = Queue.PollTaskUrls(os.Getenv("PROVISIONER_ID"), os.Getenv("WORKER_TYPE"))
+		// Set reminder to update signed urls again when they are
+		// approximately REFRESH_URLS_PREMATURELY_SECS seconds before
+		// expiring...
+		// We do this by updating updateMe channel, so that on future
+		// iterations of this select statement, we read from this new
+		// channel.
+		updateMe = time.After(signedURLs.Expires.Sub(time.Now().Add(time.Second * time.Duration(premInt))))
+		for _, q := range signedURLs.Queues {
+			log.Println("  Delete URL: " + q.SignedDeleteUrl)
+			log.Println("  Poll URL:   " + q.SignedPollUrl)
+		}
+	}
+	// Get signed urls for the first time...
+	updateUrls()
+	// loop forever, serving requests for signed urls, or requests to refresh
+	// signed urls since they are about to expire...
+	for {
```

Comment:
So you keep a fresh URL around forever... 
If you are working on a task that takes hours... This seems unnecessary...

I think you can do
```go
func (signedURLs *queue.PollTaskUrlsResponse) refresh() {
  if (signedURLs.Expires.Sub(time.Second * time.Duration(premInt)).Before(time.Now())) {
    signedURLs, _ = Queue.PollTaskUrls(os.Getenv("PROVISIONER_ID"), os.Getenv("WORKER_TYPE"))
  }
});
```

Then in `main.go` you just do:
```js
signedURLs, _ = Queue.PollTaskUrls(os.Getenv("PROVISIONER_ID"), os.Getenv("WORKER_TYPE"))
```

And before each use you just ensure that it's up to date by calling: `signedURLs.refresh()`.

**Warning,** I'm a go newbie... So I'm not sure you can just define a method for `queue.PollTaskUrlsResponse` in another module, but I suspect it just works :)

Eitherway, there is no reason to keep valid queue poll urls around when working on a task, and no need to not make that check and request synchronously.

But you probably should set `REFRESH_URLS_PREMATURELY_SECS` to 5 min to allow for slow requests and retries...